File: terminal.c
#include <emscripten.h>
#include <stdlib.h>
#include <string.h>

#define TERM_WIDTH 80
#define TERM_HEIGHT 24
#define CHAR_WIDTH 10
#define CHAR_HEIGHT 20

// Terminal state
static char terminal_buffer[TERM_HEIGHT][TERM_WIDTH];
static int cursor_x = 0;
static int cursor_y = 0;

EMSCRIPTEN_KEEPALIVE
void init_terminal() {
    // Initialize terminal buffer with spaces
    for (int y = 0; y < TERM_HEIGHT; y++) {
        for (int x = 0; x < TERM_WIDTH; x++) {
            terminal_buffer[y][x] = ' ';
        }
    }
}

EMSCRIPTEN_KEEPALIVE
void put_char(char c) {
    if (c == '\n') {
        cursor_x = 0;
        cursor_y++;
        if (cursor_y >= TERM_HEIGHT) {
            // Scroll up
            for (int y = 0; y < TERM_HEIGHT - 1; y++) {
                memcpy(terminal_buffer[y], terminal_buffer[y + 1], TERM_WIDTH);
            }
            // Clear last line
            memset(terminal_buffer[TERM_HEIGHT - 1], ' ', TERM_WIDTH);
            cursor_y = TERM_HEIGHT - 1;
        }
    } else {
        if (cursor_x >= TERM_WIDTH) {
            cursor_x = 0;
            cursor_y++;
            if (cursor_y >= TERM_HEIGHT) {
                // Scroll up
                for (int y = 0; y < TERM_HEIGHT - 1; y++) {
                    memcpy(terminal_buffer[y], terminal_buffer[y + 1], TERM_WIDTH);
                }
                // Clear last line
                memset(terminal_buffer[TERM_HEIGHT - 1], ' ', TERM_WIDTH);
                cursor_y = TERM_HEIGHT - 1;
            }
        }
        terminal_buffer[cursor_y][cursor_x] = c;
        cursor_x++;
    }
}

EMSCRIPTEN_KEEPALIVE
char* get_line(int y) {
    if (y >= 0 && y < TERM_HEIGHT) {
        return terminal_buffer[y];
    }
    return NULL;
}

EMSCRIPTEN_KEEPALIVE
int get_cursor_x() {
    return cursor_x;
}

EMSCRIPTEN_KEEPALIVE
int get_cursor_y() {
    return cursor_y;
}

EMSCRIPTEN_KEEPALIVE
void write_text(const char* text) {
    int len = strlen(text);
    for (int i = 0; i < len; i++) {
        put_char(text[i]);
    }
}

File: terminal.js
let Module = {
    onRuntimeInitialized: function() {
        initTerminal();
    }
};

const CHAR_WIDTH = 10;
const CHAR_HEIGHT = 20;
const TERM_WIDTH = 80;
const TERM_HEIGHT = 24;

let canvas, ctx;

function initTerminal() {
    canvas = document.getElementById('terminal');
    ctx = canvas.getContext('2d');
    
    // Set up canvas for terminal rendering
    ctx.font = `${CHAR_HEIGHT}px monospace`;
    ctx.textBaseline = 'top';
    ctx.fillStyle = '#00ff00'; // Classic green terminal text
    
    // Initialize the terminal buffer in WebAssembly
    Module._init_terminal();
    
    // Write some initial text
    writeToTerminal("WebAssembly Terminal v1.0\n> ");
    
    // Set up keyboard input
    document.addEventListener('keypress', handleKeyPress);
    
    // Start the render loop
    requestAnimationFrame(render);
}

function writeToTerminal(text) {
    const textPtr = Module.stringToNewUTF8(text);
    Module._write_text(textPtr);
    Module._free(textPtr);
}

function handleKeyPress(event) {
    const char = String.fromCharCode(event.charCode);
    writeToTerminal(char);
    if (event.key === 'Enter') {
        writeToTerminal('\n> ');
    }
}

function render() {
    // Clear the canvas
    ctx.fillStyle = '#000000';
    ctx.fillRect(0, 0, canvas.width, canvas.height);
    ctx.fillStyle = '#00ff00';
    
    // Render each line
    for (let y = 0; y < TERM_HEIGHT; y++) {
        const linePtr = Module._get_line(y);
        if (linePtr) {
            const line = Module.UTF8ToString(linePtr, TERM_WIDTH);
            for (let x = 0; x < TERM_WIDTH; x++) {
                ctx.fillText(
                    line[x],
                    x * CHAR_WIDTH,
                    y * CHAR_HEIGHT
                );
            }
        }
    }
    
    // Draw cursor
    const cursorX = Module._get_cursor_x();
    const cursorY = Module._get_cursor_y();
    ctx.fillStyle = '#00ff00';
    ctx.fillRect(
        cursorX * CHAR_WIDTH,
        cursorY * CHAR_HEIGHT,
        CHAR_WIDTH,
        2
    );
    
    requestAnimationFrame(render);
}
